src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs

```diff
@@ -483,32 +477,90 @@ private void EnsureCentralDirectoryRead()
 
         private void ReadCentralDirectory()
         {
+            const int ReadBufferSize = 4096;
+
+            byte[] fileBuffer = System.Buffers.ArrayPool<byte>.Shared.Rent(ReadBufferSize);
+            Span<byte> fileBufferSpan = fileBuffer.AsSpan(0, ReadBufferSize);
+
             try
             {
-                // assume ReadEndOfCentralDirectory has been called and has populated _centralDirectoryStart
+                // assume ReadEndOfCentralDirectory has been called and has populated _centralDirectoryStart and _expectedCentralDirectorySize
```

Comment:
I don't see `_expectedCentralDirectorySize` being used anywhere. It only gets values assigned to it.

Reply:
What were you intending to do with it? Does it have the purpose of just verifying that an initial valid value was already read for the central directory?

If that's the case, then I was thinking that since `_expectedCentralDirectorySize` only acquires a valid value _after_ `ReadEndOfCentralDirectory` is called, and `ReadEndOfCentralDirectory` is called in the `ZipArchive` constructor, then at the very top of the `ZipArchive` constructor, after the `ThrowIfNull` checks, you could set the initial value of `_expectedCentralDirectorySize` to `-1`:

https://github.com/dotnet/runtime/blob/e1e3cdccb872327488f8d38332f8caa540ce4d03/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L121-L123

and then here, right under this new comment, you could add a debug assert check to confirm it already acquired a valid value:

```suggestion
                // assume ReadEndOfCentralDirectory has been called and has populated _centralDirectoryStart and _expectedCentralDirectorySize
                Debug.Assert(_expectedCentralDirectorySize != -1);
```

But if that was not your intention with this field, then consider removing it.

Reply:
I should have removed `_expectedCentralDirectorySize` before submitting the PR, apologies. The original intent was twofold:
1. Use the expected size to validate the central directory as we read it. We currently validate that the file has the expected number of records in the central directory, but we don't validate the total size of the central directory.
2. Rather than performing X reads from the source stream per CD entry, consider reading `_expectedCentralDirectorySize` bytes from the stream and parsing the list of entries in memory.

I decided against doing so:
1. I wanted to avoid the breaking change for point 1.
2. Point 2 had the potential to balloon memory usage for a maliciously-constructed ZIP file. I could have overcome that by reading in 4KB chunks, but by the time I was thinking about handling central directory entries which straddled chunks, I felt that was pushing the scope of the PR.